cmd/nerdctl/network_utils.go

```diff
@@ -0,0 +1,72 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/nerdctl/pkg/labels"
+	"github.com/containerd/nerdctl/pkg/netutil/nettype"
+)
+
+func usedNetworks(ctx context.Context, containers []containerd.Container) (map[string][]string, error) {
+	used := make(map[string][]string)
+	for _, c := range containers {
+		task, err := c.Task(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		status, err := task.Status(ctx)
+		if err != nil {
+			return nil, err
+		}
+		switch status.Status {
+		case containerd.Paused, containerd.Running:
+		default:
+			continue
+		}
+		l, err := c.Labels(ctx)
+		if err != nil {
+			return nil, err
+		}
+		networkJSON, ok := l[labels.Networks]
+		if !ok {
+			continue
+		}
+		var networks []string
+		if err := json.Unmarshal([]byte(networkJSON), &networks); err != nil {
+			return nil, err
+		}
+		netType, err := nettype.Detect(networks)
+		if err != nil {
+			return nil, err
+		}
+		if netType != nettype.CNI {
```

Comment:
can we attach a network to container without implementing the cni interface ?

Reply:
Yes, see here https://github.com/containerd/nerdctl/blob/main/pkg/netutil/nettype/nettype.go#L27-L31